app/org/sagebionetworks/bridge/services/SchedulePlanServiceImpl.java

```diff
@@ -58,5 +75,38 @@ public SchedulePlan updateSchedulePlan(SchedulePlan plan) {
     public void deleteSchedulePlan(StudyIdentifier studyIdentifier, String guid) {
         schedulePlanDao.deleteSchedulePlan(studyIdentifier, guid);
     }
-    
+
+    /**
+     * If the activity has a survey reference, look up the survey's identifier. Don't trust the client to 
+     * supply the correct one for the survey's primary keys. We're adding this when writing schedules because 
+     * the clients have come to depend on it, and this is more efficient than looking it up on every read.
+     * 
+     * @param studyId
+     * @param activity
+     * @return
+     */
+    private void lookupSurveyReferenceIdentifiers(StudyIdentifier studyId, SchedulePlan plan) {
+        for (Schedule schedule : plan.getStrategy().getAllPossibleSchedules()) {
+            for (int i=0; i < schedule.getActivities().size(); i++) {
+                Activity activity = schedule.getActivities().get(i);
+                activity = updateActivityWithSurveyIdentifier(studyId, activity);
+                schedule.getActivities().set(i, activity);
+            }
+        }
+    }
+
+    private Activity updateActivityWithSurveyIdentifier(StudyIdentifier studyId, Activity activity) {
+        if (activity.getSurvey() != null) {
+            SurveyReference ref = activity.getSurvey();
+            Survey survey = surveyService.getSurveyMostRecentlyPublishedVersion(studyId, ref.getGuid());
```

Comment:
IIRC, researchers can change the identifier between survey versions (although one could argue they shouldn't). If we always get the most recently published, this could potentially return the wrong identifier. Would it be better to have the getMostRecentlyPublished call when createdOn is null and getSurvey when createdOn is not null?